californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -139,39 +159,38 @@ public BlockwiseLayer(final NetworkConfig config) {
 	@Override
 	public void sendRequest(final Exchange exchange, final Request request) {
 
-		BlockOption block2 = request.getOptions().getBlock2();
-		if (block2 != null && block2.getNum() > 0) {
-			// This is the case if the user has explicitly added a block option
-			// for random access.
-			// Note: We do not regard it as random access when the block num is
-			// 0. This is because the user might just want to do early block
-			// size negotiation but actually wants to receive all blocks.
-			LOGGER.fine("request contains block2 option, creating random-access blockwise status");
-			BlockwiseStatus status = new BlockwiseStatus(getSizeForSzx(block2.getSzx()), request.getOptions().getContentFormat());
-			status.setCurrentSzx(block2.getSzx());
-			status.setCurrentNum(block2.getNum());
-			status.setRandomAccess(true);
-			exchange.setResponseBlockStatus(status);
-			lower().sendRequest(exchange, request);
-
-		} else if (requiresBlockwise(request)) {
-			// This must be a large POST or PUT request
-			startBlockwiseUpload(exchange, request);
+		Request requestToSend = request;
 
-		} else {
-			// no blockwise transfer required
-			exchange.setCurrentRequest(request);
-			lower().sendRequest(exchange, request);
+		if (isTransparentBlockwiseHandlingEnabled()) {
+
+			BlockOption block2 = request.getOptions().getBlock2();
+			if (block2 != null && block2.getNum() > 0) {
+				// This is the case if the user has explicitly added a block option
+				// for random access.
+				// Note: We do not regard it as random access when the block number is 0.
+				// This is because the user might just want to do early block
+				// size negotiation but actually want to retrieve the whole body by means of
+				// a transparent blockwise transfer.
+				LOGGER.fine("outbound request contains block2 option, creating random-access blockwise status");
+				addRandomAccessBlock2Status(exchange, request, block2);
+
+			} else if (requiresBlockwise(request)) {
+				// This must be a large POST or PUT request
+				requestToSend = startBlockwiseUpload(exchange, request);
+
+			}
 		}
+
+		exchange.setCurrentRequest(requestToSend);
+		lower().sendRequest(exchange, requestToSend);
 	}
 
-	private void startBlockwiseUpload(final Exchange exchange, final Request request) {
+	private Request startBlockwiseUpload(final Exchange exchange, final Request request) {
 
-		BlockwiseStatus status = findRequestBlockStatus(exchange, request);
+		final KeyUri key = getKey(exchange, request);
+		Block1BlockwiseStatus status = getOutboundBlock1Status(key, exchange, request);
```

Comment:
My analogy just aims to explain why this could be considered as a deliberated action, nothing more. All of this depends on how you describe your API. This was just in response of :

>So far, canceling was something the client code had to do deliberately, wasn't it?

(I should be more cautious with analogies :/ )